src/Framework/Constraint/Constraint.php

```diff
@@ -152,4 +152,52 @@ protected function failureDescription($other): string
     {
         return $this->exporter()->export($other) . ' ' . $this->toString();
     }
+
+    /**
+     * Returns a custom string representation of the constraint object when it
+     * appears in context of an $operator expression
+     *
+     * The purpose of this method is to provide meaningful messages in context
+     * of operators such as LogicalNot. Native PHPUnit constraints are supported
+     * out of the box by LogicalNot, but externally developed ones had no way to
+     * provide correct messages in this context.
+     *
+     * The method shall return null, if there is no need for customization in the
+     * context of given $operator. The $position starts with 0 for connective
+     * operators, but for unary operator, such as "not" operator, it equals 1 to
+     * denote that this constraint is on the right hand side of the $operator.
+     *
+     * @param Operator $operator the $operator of the expression
+     * @param int      $position position in $operator expression
+     */
+    protected function toStringInContext(Operator $operator, int $position): ?string
```

Comment:
I interpret null as "not implemented, please use toString() instead" in several places. Generally, empty string is a special case of the description string. Maybe there is another way to pass this information.

Reply:
Would raising an exception make sense when the `Constraint` cannot be represented as a string?

Reply:
Well, then the default behavior will be to raise exception. Maybe my wording was quite misleading, by "not implemented" I meant that the sting is not specialized for the context of given $operator and should be taken from ``toString()`` instead. I'll see if it's possible to refactor it such that the function fallbacks to ``toString()`` by itself. If it's possible, then it will always return string.